test/functional/options/defaults_spec.lua

```diff
@@ -380,3 +382,279 @@ describe('XDG-based defaults', function()
+      it('knows XDG_DATA_HOME', function()
+        clear({env={
+          XDG_DATA_HOME=alter_path('/home/docwhat/.local'),
+        }})
+        eq(alter_path('/home/docwhat/.local' .. appended_dir), funcs.stdpath('data'))
+      end)
+
+      it('handles changes during runtime', function()
+        clear({env={
+          XDG_DATA_HOME=alter_path('/home/original'),
+        }})
+        eq(alter_path('/home/original' .. appended_dir), funcs.stdpath('data'))
+        command('let $XDG_DATA_HOME="/home/new"')
+        eq(alter_path('/home/new' .. appended_dir), funcs.stdpath('data'))
+      end)
+
+      it("doesn't expand $VARIABLES", function()
+        clear({env={
+          XDG_DATA_HOME='$VARIABLES',
+          VARIABLES='this-should-not-happen',
+        }})
+        eq(alter_path('$VARIABLES' .. appended_dir), funcs.stdpath('data'))
+      end)
+
+      it("doesn't expand ~/", function()
+        clear({env={
+          XDG_DATA_HOME='~/frobnitz',
+        }})
+        eq(alter_path('~/frobnitz' .. appended_dir), funcs.stdpath('data'))
+      end)
+    end)
+
+    describe('with "cache"' , function ()
+      it('knows XDG_CACHE_HOME', function()
+        clear({env={
+          XDG_CACHE_HOME=alter_path('/home/docwhat/.cache'),
+        }})
+        eq(alter_path('/home/docwhat/.cache/nvim'), funcs.stdpath('cache'))
+      end)
+
+      it('handles changes during runtime', function()
+        clear({env={
+          XDG_CACHE_HOME=alter_path('/home/original'),
+        }})
+        eq(alter_path('/home/original/nvim'), funcs.stdpath('cache'))
+        command('let $XDG_CACHE_HOME="/home/new"')
+        eq(alter_path('/home/new/nvim'), funcs.stdpath('cache'))
+      end)
+
+      it("doesn't expand $VARIABLES", function()
+        clear({env={
+          XDG_CACHE_HOME='$VARIABLES',
+          VARIABLES='this-should-not-happen',
+        }})
+        eq(alter_path('$VARIABLES/nvim'), funcs.stdpath('cache'))
+      end)
+
+      it("doesn't expand ~/", function()
+        clear({env={
+          XDG_CACHE_HOME='~/frobnitz',
+        }})
+        eq(alter_path('~/frobnitz/nvim'), funcs.stdpath('cache'))
+      end)
+    end)
+  end)
+
+  context('returns a List', function()
+    -- Some OS specific variables the system would have set.
+    local function base_env()
+      if helpers.iswin() then
+        return {
+          HOME='C:\\Users\\docwhat', -- technically, is not a usual PATH
+          HOMEDRIVE='C:',
+          HOMEPATH='\\Users\\docwhat',
+          LOCALAPPDATA='C:\\Users\\docwhat\\AppData\\Local',
+          TEMP='C:\\Users\\docwhat\\AppData\\Local\\Temp',
+          TMP='C:\\Users\\docwhat\\AppData\\Local\\Temp',
+        }
+      else
+        return {
+          HOME='/home/docwhat',
+          HOMEDRIVE='HOMEDRIVE-should-be-ignored',
+          HOMEPATH='HOMEPATH-should-be-ignored',
+          LOCALAPPDATA='LOCALAPPDATA-should-be-ignored',
+          TEMP='TEMP-should-be-ignored',
+          TMP='TMP-should-be-ignored',
```

Comment:
There also is `$TMPDIR`, used by `mktemp`. Should be ignored as well, XDG spec does not say anything about it.